core/iwasm/libraries/lib-wasi-threads/test/common.h

```diff
@@ -6,11 +6,16 @@
 #include <stdlib.h>
 #include <stdio.h>
 #include <assert.h>
-#include <pthread.h>
 #include <stdbool.h>
 #include <unistd.h>
 #include <limits.h>
 
+#if USE_PTHREAD_SYNC_PRIMITIVES != 0
+#include <pthread.h>
+#else
+#include "sync_primitives.h"
```

Comment:
Yes, had better use the `pthread_barrier_wait` after the hang issue is resolved. But note that there will be some normal load/store data races in wasi-libc implementation, e.g. a_swap:
https://github.com/WebAssembly/wasi-libc/blob/main/libc-top-half/musl/src/internal/atomic.h#L106-L115
```C
	int old;
	do old = *p;
	while (a_cas(p, old, v) != old);
	return old;
```
Here `old = *p` is not atomic, and a_cas(p, old, v) is atomic, if two threads operate on the same address `p`, data races may be reported by sanitizer.

Reply:
> Yes, had better use the pthread_barrier_wait after the hang issue is resolved.

So should we merge this one as it is then? Until we get it fixed in wasi-libc.

> Here old = *p is not atomic, and a_cas(p, old, v) is atomic, if two threads operate on the same address p, data races may be reported by sanitizer.

Nice catch, that's probably what the thread sanitizer is complaining about for the load/store data races. It may be the same for other operations as well, I see the `a_fetch_add` a few lines later doing something similar.